File: packages/mcp-bridge-client-py/mcp_bridge_client/tool_catalog.py
"""Tool catalog with caching support"""

from typing import Dict, List, Optional
from time import time


class ToolCatalog:
    """Tool catalog with caching support"""
    
    def __init__(self, ttl: int = 300):
        """
        Initialize tool catalog
        
        Args:
            ttl: Cache time-to-live in seconds
        """
        self.tools: List = []
        self.last_update: float = 0
        self.ttl = ttl
    
    def set_all(self, tools: List) -> None:
        """Set all tools in catalog"""
        self.tools = tools
        self.last_update = time()
    
    def get_all(self) -> List:
        """
        Get all tools from catalog
        Returns empty list if cache expired
        """
        if self.is_expired():
            return []
        return self.tools
    
    def is_expired(self) -> bool:
        """Check if cache is expired"""
        return time() - self.last_update > self.ttl
    
    def clear(self) -> None:
        """Clear catalog"""
        self.tools = []
        self.last_update = 0
    
    def get_by_server(self, server_name: str) -> List:
        """Get tools by server"""
        return [t for t in self.tools if t.server == server_name]
    
    def get_by_name(self, tool_name: str) -> Optional[object]:
        """Get tool by name"""
        for tool in self.tools:
            if tool.name == tool_name:
                return tool
        return None
    
    def get_stats(self) -> Dict[str, any]:
        """Get catalog statistics"""
        servers = set(t.server for t in self.tools)
        return {
            'total': len(self.tools),
            'servers': len(servers),
            'last_update': self.last_update,
            'expired': self.is_expired()
        }

File: packages/mcp-bridge-client/src/toolCatalog.ts
import { MCPTool } from './client';

/**
 * Tool catalog with caching support
 */
export class ToolCatalog {
    private tools: MCPTool[] = [];
    private lastUpdate: number = 0;
    private ttl: number;

    constructor(ttl: number = 300000) {
        this.ttl = ttl;
    }

    /**
     * Set all tools in catalog
     */
    setAll(tools: MCPTool[]): void {
        this.tools = tools;
        this.lastUpdate = Date.now();
    }

    /**
     * Get all tools from catalog
     * Returns empty array if cache expired
     */
    getAll(): MCPTool[] {
        if (this.isExpired()) {
            return [];
        }
        return this.tools;
    }

    /**
     * Check if cache is expired
     */
    isExpired(): boolean {
        return Date.now() - this.lastUpdate > this.ttl;
    }

    /**
     * Clear catalog
     */
    clear(): void {
        this.tools = [];
        this.lastUpdate = 0;
    }

    /**
     * Get tools by server
     */
    getByServer(serverName: string): MCPTool[] {
        return this.tools.filter(t => t.server === serverName);
    }

    /**
     * Get tool by name
     */
    getByName(toolName: string): MCPTool | undefined {
        return this.tools.find(t => t.name === toolName);
    }

    /**
     * Get catalog statistics
     */
    getStats(): { total: number; servers: number; lastUpdate: number; expired: boolean } {
        const servers = new Set(this.tools.map(t => t.server));
        return {
            total: this.tools.length,
            servers: servers.size,
            lastUpdate: this.lastUpdate,
            expired: this.isExpired()
        };
    }
}
